src/messages/mod.rs

```diff
@@ -232,76 +305,80 @@ impl SignedRoutingMessage {
-        let invalid = self
-            .signatures
-            .sigs
-            .iter()
-            .filter_map(|(pub_id, sig)| {
-                // Remove if not in sending nodes or signature is invalid:
-                let is_valid = if let Authority::Client { ref client_id, .. } = self.content.src {
-                    client_id == pub_id
-                        && client_id
-                            .signing_public_key()
-                            .verify_detached(sig, &signed_bytes)
-                } else {
-                    self.is_sender(pub_id)
-                        && pub_id
-                            .signing_public_key()
-                            .verify_detached(sig, &signed_bytes)
-                };
-                if is_valid {
-                    None
-                } else {
-                    Some(*pub_id)
+    fn find_invalid_sigs(&self, signed_bytes: &[u8]) -> Vec<BlsPublicKeyShare> {
+        match self.security_metadata {
+            SecurityMetadata::None | SecurityMetadata::Full(_) => vec![],
+            SecurityMetadata::Partial(ref partial) => {
+                let invalid: Vec<_> = partial
+                    .shares
+                    .iter()
+                    .filter(|&(key, sig)| !key.verify(sig, signed_bytes))
+                    .map(|(key, _)| *key)
+                    .collect();
+                if !invalid.is_empty() {
+                    debug!("{:?}: invalid signatures: {:?}", self, invalid);
                 }
-            })
-            .collect_vec();
-        if !invalid.is_empty() {
-            debug!("{:?}: invalid signatures: {:?}", self, invalid);
+                invalid
+            }
         }
-        invalid
     }
 
     // Returns true if there are enough signatures (note that this method does not verify the
     // signatures, it only counts them; it also does not verify `self.src_section`).
     fn has_enough_sigs(&self) -> bool {
-        use crate::Authority::*;
-
         // Only Clients are allowed to omit the src_section
         if !self.content.src.is_client() && self.src_section.is_none() {
             return false;
         }
 
-        match self.content.src {
-            ClientManager(_) | NaeManager(_) | NodeManager(_) | Section(_) | PrefixSection(_) => {
-                let valid_sigs = self.signatures.len();
-                valid_sigs * QUORUM_DENOMINATOR > self.src_size() * QUORUM_NUMERATOR
-            }
-            ManagedNode(_) | Client { .. } => self.signatures.len() == 1,
+        match &self.security_metadata {
+            SecurityMetadata::None => !self.content.src.is_multiple(),
+            SecurityMetadata::Partial(partial) => partial.shares.len() > partial.pk_set.threshold(),
+            SecurityMetadata::Full(_) => true,
+        }
+    }
+
+    #[cfg(test)]
+    pub fn signatures(&self) -> Option<&BTreeMap<BlsPublicKeyShare, BlsSignatureShare>> {
```

Comment:
better have the name as `partial_signatures` to avoid mis-use?

Reply:
It is a test function and the return type is very specific, can probably wait to be cleaned up later.